### PR TITLE
Lib modified

### PR DESCRIPTION
**Pull request by hafsaanis123**

### CODE REVIEW

File: src/app.module.ts [hunk 1]
```diff
@@ -1,25 +1,69 @@
...
     AssetFundamentalsModule,
+    AssetSegmentsModule,
+    //AssetEssentialsModule,
+    AssetMetricsModule,
+    AssetEssentialsRtModule,
+    AssetEssentialsWrtModule,
   ],
-  controllers: [AppController],
+  controllers: [AppController, AssetDetailsController,AssetEssentialsController, AssetFundamentalsController, AssetSegmentsController, AssetMetricsController, AssetEssentialsRtController, AssetEssentialsWrtController],
+  providers: [AssetDetailsService, AssetEssentialsService, AssetFundamentalsService, AssetSegmentsService, AssetMetricsService, AssetEssentialsRtService, AssetEssentialsWrtService],
```
Comment (OsamaQureshi147) [inline src/app.module.ts]: I think we should revert the changes on this file. We don't have to import all the modules, controllers, entities and services here. Just the modules are enough in imports section. Also, here just to instantiate a DB is enough, we need to declare the TypeORM entities in their respective modules like `TypeOrmModule.forFeature([xyz.entity.ts])`.
Let me know if there is still an issue.

File: src/app.controller.ts [hunk 2]
```diff
@@ -1,3 +1,4 @@
+import 'reflect-metadata';
```
Comment (OsamaQureshi147) [inline src/app.controller.ts]: why we need to import this here?

File: src/modules/asset-metrics/asset-metrics.controller.spec.ts [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+      controllers: [AssetMetricsController],
+    }).compile();
+
+    controller = module.get<AssetMetricsController>(AssetMetricsController);
+  });
+
+  it('should be defined', () => {
+    expect(controller).toBeDefined();
+  });
+});
```
Comment (OsamaQureshi147) [inline src/modules/asset-metrics/asset-metrics.controller.spec.ts]: Delete all `.spec` files for now. We are not doing unit testing for now.

File: src/modules/asset-segments/asset-segments.service.ts [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+    @InjectRepository(AssetSegmentsEntity)
+    private readonly assetSegmentsRepository: Repository<AssetSegmentsEntity>,
+  ) {}
+
+  async create(assetSegmentsDto: AssetSegmentsDto): Promise<AssetSegmentsEntity> {
+    const assetSegment = this.assetSegmentsRepository.create(assetSegmentsDto);
+    return this.assetSegmentsRepository.save(assetSegment);
+  }
+
+  async findAll(query: any): Promise<AssetSegmentsEntity[]> {
```
Comment (OsamaQureshi147) [inline src/modules/asset-segments/asset-segments.service.ts]: Always avoid using `any` in typescript. Also, did you get an error on linting or pre-commit hook here? I am asking because if you didn't get any then I need to fix this on your system as well.

File: src/modules/asset-segments/asset-segments.service.spec.ts [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+      providers: [AssetSegmentsService],
+    }).compile();
+
+    service = module.get<AssetSegmentsService>(AssetSegmentsService);
+  });
+
+  it('should be defined', () => {
+    expect(service).toBeDefined();
+  });
+});
```
Comment (OsamaQureshi147) [inline src/modules/asset-segments/asset-segments.service.spec.ts]: delete this spec file too

File: src/modules/asset-segments/asset-segments.module.ts [hunk 6]
```diff
@@ -1,4 +1,15 @@
...
+import { AssetSegmentsEntity } from 'lib-typeorm';
 
-@Module({})
+@Module({
+  imports: [
+    TypeOrmModule.forFeature([AssetSegmentsEntity]),
+  ],
+  controllers: [AssetSegmentsController],
+  providers: [AssetSegmentsService],
+  exports: [TypeOrmModule],
```
Comment (OsamaQureshi147) [inline src/modules/asset-segments/asset-segments.module.ts]: why we are exporting `TypeOrmModule` here?

File: src/modules/asset-segments/asset-segments.controller.ts [hunk 7]
```diff
@@ -1,4 +1,35 @@
...
+export class AssetSegmentsController {
+  constructor(private readonly assetSegmentsService: AssetSegmentsService) {}
+
+  @Post()
+  async create(@Body() assetSegmentsDto: AssetSegmentsDto): Promise<AssetSegmentsEntity> {
+    return this.assetSegmentsService.create(assetSegmentsDto);
+  }
+
+  @Get()
+  async findAll(@Query() query: any): Promise<AssetSegmentsEntity[]> {
```
Comment (OsamaQureshi147) [inline src/modules/asset-segments/asset-segments.controller.ts]: Never use `any` in typescript.

File: src/modules/asset-segments/asset-segments.service.ts [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+      queryBuilder.andWhere('asset_segments.metric = :metric', { metric: query.metric });
+    }
+
+    if (query.segmentType) {
+      queryBuilder.andWhere('asset_segments.segmentType = :segmentType', { segmentType: query.segmentType });
+    }
+
+    if (query.periodicity) {
+      queryBuilder.andWhere('asset_segments.periodicity = :periodicity', { periodicity: query.periodicity });
+    }
```
Comment (OsamaQureshi147) [inline src/modules/asset-segments/asset-segments.service.ts]: why we are using querybuilder here with so many if statements? Is it possible to do this with TypeORM query? Let me know if I am missing something.

File: src/modules/asset-metrics/asset-metrics.service.ts [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+    }
+
+    if (query.subCategory) {
+      queryBuilder.andWhere('asset_metrics.subCategory = :subCategory', { subCategory : `%${query.subcategory}%` });
+    }
+
+    if (query.dataType) {
+      queryBuilder.andWhere('asset_metrics.dataType = :dataType', { dataType : `%${query.dataType}%` });
+    }
+
```
Comment (OsamaQureshi147) [inline src/modules/asset-metrics/asset-metrics.service.ts]: same remark as below. Why we are using querybuilder with so many if statements?

File: src/modules/asset-fundamentals/asset-fundamentals.service.ts [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+    }
+
+    if (query.epsReportDate) {
+      queryBuilder.andWhere('asset_fundamentals.epsReportDate = :epsReportDate', { epsReportDate: query.epsReportDate });
+    }
+
+    if (query.value) {
+      queryBuilder.andWhere('asset_fundamentals.value = :value', { value: query.value });
+    }
+
```
Comment (OsamaQureshi147) [inline src/modules/asset-fundamentals/asset-fundamentals.service.ts]: why we are using query builder here?

File: src/modules/asset-metrics/asset-metrics.module.ts [hunk 11]
```diff
@@ -0,0 +1,17 @@
...
+import { AssetFundamentalsService } from '../asset-fundamentals/asset-fundamentals.service';
+
+
+@Module({
+  imports: [
+    TypeOrmModule.forFeature([AssetFundamentalsEntity, AssetMetricsEntity]),
+  ],
+  providers: [AssetFundamentalsService],
+  controllers: [AssetFundamentalsController],
+  exports: [TypeOrmModule],
```
Comment (OsamaQureshi147) [inline src/modules/asset-metrics/asset-metrics.module.ts]: why we are exporting typeorm module

File: src/modules/asset-essentials-rt/asset-essentials-rt.service.spec.ts [hunk 12]
```diff
@@ -0,0 +1,18 @@
...
+      providers: [AssetEssentialsRtService],
+    }).compile();
+
+    service = module.get<AssetEssentialsRtService>(AssetEssentialsRtService);
+  });
+
+  it('should be defined', () => {
+    expect(service).toBeDefined();
+  });
+});
```
Comment (OsamaQureshi147) [inline src/modules/asset-essentials-rt/asset-essentials-rt.service.spec.ts]: remove this file

File: src/modules/asset-essentials-rt/asset-essentials-rt.service.ts [hunk 13]
```diff
@@ -0,0 +1,98 @@
...
+      queryBuilder.andWhere('asset_essentials_rt.volume = :volume', { volume: query.volume });
+    }
+
+    if (query.changePercent) {
+      queryBuilder.andWhere('asset_essentials_rt.changePercent = :changePercent', { changePercent: query.changePercent });
+    }
+
+    if (query.marketCap) {
+      query.Builder.andWhere('asset_essentials_rt.marketCap = :marketCap', {marketCap: query.marketCap})
+    }
```
Comment (OsamaQureshi147) [inline src/modules/asset-essentials-rt/asset-essentials-rt.service.ts]: same as below. WHy we are using query builder

**Review by OsamaQureshi147 — CHANGES_REQUESTED** [review]

**Comment by OsamaQureshi147** [comment]
Good work @hafsaanis123. A few things to look at. Kindly provide reasoning for some concerns I have raised. I am new to NestJS too so my remarks could be invalid at some points.

File: src/modules/asset-essentials-rt/asset-essentials-rt.service.ts (hunk 13 above)
Comment (hafsaanis123) [reply src/modules/asset-essentials-rt/asset-essentials-rt.service.ts]: Used an alternative method now. Now i am using an array that stores the conditions for the queries

File: src/modules/asset-essentials-rt/asset-essentials-rt.service.spec.ts (hunk 12 above)
Comment (hafsaanis123) [reply src/modules/asset-essentials-rt/asset-essentials-rt.service.spec.ts]: done

File: src/modules/asset-fundamentals/asset-fundamentals.service.ts (hunk 10 above)
Comment (hafsaanis123) [reply src/modules/asset-fundamentals/asset-fundamentals.service.ts]: Used an alternative method now. Now i am using an array that stores the conditions for the queries

File: src/modules/asset-metrics/asset-metrics.module.ts (hunk 11 above)
Comment (hafsaanis123) [reply src/modules/asset-metrics/asset-metrics.module.ts]: this was to be done in the library instead. have made the changes in the library

File: src/modules/asset-metrics/asset-metrics.service.ts (hunk 9 above)
Comment (hafsaanis123) [reply src/modules/asset-metrics/asset-metrics.service.ts]: Used an alternative method now. Now i am using an array that stores the conditions for the queries

File: src/modules/asset-segments/asset-segments.service.ts (hunk 8 above)
Comment (hafsaanis123) [reply src/modules/asset-segments/asset-segments.service.ts]: Used an alternative method now. Now i am using an array that stores the conditions to build dynamic queries

File: src/modules/asset-segments/asset-segments.controller.ts (hunk 7 above)
Comment (hafsaanis123) [reply src/modules/asset-segments/asset-segments.controller.ts]: fixed

File: src/modules/asset-segments/asset-segments.module.ts (hunk 6 above)
Comment (hafsaanis123) [reply src/modules/asset-segments/asset-segments.module.ts]: Had to do this in the library instead. fixed

File: src/modules/asset-segments/asset-segments.service.spec.ts (hunk 5 above)
Comment (hafsaanis123) [reply src/modules/asset-segments/asset-segments.service.spec.ts]: done

File: src/modules/asset-segments/asset-segments.service.ts (hunk 4 above)
Comment (hafsaanis123) [reply src/modules/asset-segments/asset-segments.service.ts]: I didnt get any error
